### ISSUE TITLE
UIParticleSystem does not work on Build

### ISSUE THREAD

**Issue by SimonDarksideJ** [Unity-UI-Extensions/com.unity.uiextensions#198]
<sup>Issue created by &lt;unknown&gt; as [Bitbucket Issue #&#8203;162](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162) on 2017.08.18 15:31.</sup>
The script works fine in the editor but not on windows or android build. It used to work fine with previous versions of the asset.

Here is the error that I get during runtime (using Unity 2017.1 (64bit). I created an empty project and imported UnityUIExtensions):

**NullReferenceException
  at (wrapper managed-to-native) UnityEngine.Material:Internal_CreateWithShader (UnityEngine.Material,UnityEngine.Shader)
  at UnityEngine.Material..ctor (UnityEngine.Shader shader) [0x00009] in C:\buildslave\unity\build\artifacts\generated\common\runtime\ShaderBindings.gen.cs:260 
  at UnityEngine.UI.Extensions.UIParticleSystem.Initialize () [0x000d1] in C:\Unity\PARTICLETEST\Assets\unity-ui-extensions\Scripts\Effects\UIParticleSystem.cs:74 
  at UnityEngine.UI.Extensions.UIParticleSystem.Awake () [0x00008] in C:\Unity\PARTICLETEST\Assets\unity-ui-extensions\Scripts\Effects\UIParticleSystem.cs:122 
 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/ShaderBindings.gen.cs Line: 260)**

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.18 16:21](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-39177477), @SimonDarksideJ commented:</sup>
That is so random. I did Test this before merge in several builds, inc webgl which is really tight. 
Will investigate.  Are you using:
* source
* 1.x build
* new 2.x build

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.19 14:11](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-39190091), FireMouse commented:</sup>
Thanks for the quick reply.

I'm using the latest 2.x version that I got from the downloads section a couple of days ago. I had not updated the asset for months before that. Perhaps I should also add that I'm creating particles on a Canvas with Screen Space - Overlay render mode. I don't change any settings when I add UIParticleSystem on it.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.23 04:36](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-39256492), cAptAin rAinmAker commented:</sup>
I faced the problem too. Can not make UI ParticleSystem run as expected on Android build.
In logcat it said something null in shader.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.28 08:33](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-39430308), cAptAin rAinmAker commented:</sup>
@FireMouse which version of the asset your worked fine?

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.11 11:56](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-40457228), B.A. Rops commented:</sup>
Same issue here, Unity 5.6.3p4 on windows phone 8.1 builds

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.20 17:25](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-40681067), OddButAwesome commented:</sup>
Same issue. Unity 5.5.0f3. UIExt downloaded 21/10/17. 

Particles visible in editor and in-Unity Game window but not on Android devices (tabs/phones). No errors/msgs in Unity log window. 

Tried different shaders/materials (included UIExtensions and Unity std). Canvas sort orders. Cam culling.  Canvas-Screenspace overlay. Scale with screen size.  

Hope you solve it soon. Appreciate your efforts in putting it all together in the first place BTW.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.24 15:58](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-40752696), Nathan van Hulst commented:</sup>
Same issue on 5.6.1f1, UIParticles do not appear in build, in editor everything is fine.
Too bad as I really wanted to use this but now I can't.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.24 19:58](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-40757744), Circaetus commented:</sup>
Unfortunately it doesn't work for me, I use Unity 5.6.3p4 and built for iOS with the result that I don't see ParticlSystem, Xcode delivers the next problem:

NullReferenceException: A null value was found where an object instance was required.
  at UnityEngine.UI.Extensions.UIParticleSystem.Initialize () [0x00000] in <filename unknown>:0 
  at UnityEngine.UI.Extensions.UIParticleSystem.Awake () [0x00000] in <filename unknown>:0 
 
(Filename: currently not available on il2cpp Line: -1)

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.27 11:40](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-40832407), @SimonDarksideJ commented:</sup>
Check the UI Forum where this script was originally announced.

https://forum.unity.com/threads/free-script-particle-systems-in-ui-screen-space-overlay.406862/#post-3211269

You need to make sure you are using the shader included within the script for platforms or it won't render.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.30 14:57](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-40893212), OddButAwesome commented:</sup>
Simon you legend. 

First test seems to work.  It appears it was shader selection. I was using "UI Extensions/Particles/Additive (Soft)" where the code looks for "UI Extensions/Particles/Additive". Looking at the script, line 73 actually says so... doh. Good thing I RTFM  ><.  

Hopefully trails will get fixed soon. 

Will report if there are any further issues.

Thanks heaps.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.09.29 05:35](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/162#comment-48073309), Former user commented:</sup>
The code for this script could be improved.
In the Initialize method, whether a material has been assigned to the script or not, a new material will be created using Shader.Find() (which is not a good idea), but that newly created material will only be assigned if the material assigned in the inspector is null.
In short, it's like this:

var shader = Shader.Find(...)
var pmaterial = new Material(shader);
if (material == null) material = pmaterial

Which is rather silly. I refactored it to:

if (material == null) 
{
    var shader = Shader.Find(...)
    var pmaterial = new Material(shader);
    material = pmaterial
}
